File: orig_app/internal/handler/google_api/authorizations_controller.go
package googleapi

import (
	"fmt"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/pkg/errors"
)

// AuthorizationsController handles Google API authorization callbacks
type AuthorizationsController struct {
	cloudPlatformClient *CloudPlatformClient
}

// NewAuthorizationsController creates a new AuthorizationsController
func NewAuthorizationsController(cloudPlatformClient *CloudPlatformClient) *AuthorizationsController {
	return &AuthorizationsController{
		cloudPlatformClient: cloudPlatformClient,
	}
}

// Callback handles the response from Google after the user
// goes through authentication and authorization process
func (c *AuthorizationsController) Callback(ctx *gin.Context) {
	// Get redirect URI from session
	redirectURI, err := c.getRedirectURIFromSession(ctx)
	if err != nil {
		ctx.JSON(http.StatusForbidden, gin.H{"error": "Access denied"})
		return
	}

	// Check if there's an error parameter (user declined authorization)
	if errorParam := ctx.Query("error"); errorParam != "" {
		ctx.SetFlash("alert", "Google Cloud authorizations required")
		// Get error URI from session
		errorURI, _ := ctx.Get("error_uri")
		if errorURI != nil {
			redirectURI = errorURI.(string)
		}
		ctx.Redirect(http.StatusFound, redirectURI)
		return
	}

	// Check if there's a code parameter (successful authorization)
	if code := ctx.Query("code"); code != "" {
		// Get callback URL for Google API
		callbackURL := fmt.Sprintf("%s/google_api/auth/callback", ctx.Request.Host)

		// Get token from Google API
		token, expiresAt, err := c.cloudPlatformClient.GetToken(code, callbackURL)
		if err != nil {
			// Handle timeout or connection errors
			if errors.Is(err, ErrTimeout) || errors.Is(err, ErrConnectionFailed) {
				ctx.SetFlash("alert", "Timeout connecting to the Google API. Please try again.")
				ctx.Redirect(http.StatusFound, redirectURI)
				return
			}
			// Handle other errors
			ctx.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}

		// Store token and expiration in session
		state := ctx.Query("state")
		if state != "" {
			ctx.Set(CloudPlatformClient.SessionKeyForToken(state), token)
			ctx.Set(CloudPlatformClient.SessionKeyForExpiresAt(state), expiresAt.Format(time.RFC3339))
		}
	}

	// Redirect to the appropriate URI
	ctx.Redirect(http.StatusFound, redirectURI)
}

// getRedirectURIFromSession retrieves the redirect URI from the session
func (c *AuthorizationsController) getRedirectURIFromSession(ctx *gin.Context) (string, error) {
	state := ctx.Query("state")
	if state == "" {
		return "", fmt.Errorf("state parameter is required")
	}

	sessionKey := CloudPlatformClient.SessionKeyForRedirectURI(state)
	redirectURI, exists := ctx.Get(sessionKey)
	if !exists {
		return "", fmt.Errorf("redirect URI not found in session")
	}

	return redirectURI.(string), nil
}

// CloudPlatformClient represents a client for Google Cloud Platform API
type CloudPlatformClient struct {
	// Add fields as needed
}

// NewCloudPlatformClient creates a new CloudPlatformClient
func NewCloudPlatformClient() *CloudPlatformClient {
	return &CloudPlatformClient{}
}

// GetToken retrieves a token from Google API
func (c *CloudPlatformClient) GetToken(code, redirectURI string) (string, time.Time, error) {
	// Implementation would go here
	// This is a placeholder for the actual implementation
	return "token", time.Now().Add(1 * time.Hour), nil
}

// SessionKeyForToken returns the session key for storing the token
func (c *CloudPlatformClient) SessionKeyForToken(state string) string {
	return fmt.Sprintf("google_api_token_%s", state)
}

// SessionKeyForExpiresAt returns the session key for storing the token expiration
func (c *CloudPlatformClient) SessionKeyForExpiresAt(state string) string {
	return fmt.Sprintf("google_api_expires_at_%s", state)
}

// SessionKeyForRedirectURI returns the session key for storing the redirect URI
func (c *CloudPlatformClient) SessionKeyForRedirectURI(state string) string {
	return fmt.Sprintf("google_api_redirect_uri_%s", state)
}

// Error types
var (
	ErrTimeout           = errors.New("timeout connecting to Google API")
	ErrConnectionFailed  = errors.New("connection to Google API failed")
)

File: rust_app/src/controllers/google_api/authorizations_controller.rs
// Ported from orig_app/app/controllers/google_api/authorizations_controller.rb
// Date ported: 2025-04-30

use actix_web::{get, post, web, HttpRequest, HttpResponse, Responder};
use serde::Deserialize;

// Placeholder for session and flash handling
type Session = std::collections::HashMap<String, String>;

// Placeholder for GoogleApi::CloudPlatform::Client logic
struct GoogleApiCloudPlatformClient;

impl GoogleApiCloudPlatformClient {
    fn new(_user: Option<&str>, _callback_url: &str) -> Self {
        Self {}
    }
    fn get_token(&self, _code: &str) -> (String, String) {
        // Simulate token and expiry
        ("token_value".to_string(), "expires_at_value".to_string())
    }
    fn session_key_for_token() -> &'static str {
        "google_api_token"
    }
    fn session_key_for_expires_at() -> &'static str {
        "google_api_expires_at"
    }
    fn session_key_for_redirect_uri(_state: &str) -> String {
        format!("google_api_redirect_uri_{}", _state)
    }
}

#[derive(Deserialize)]
pub struct CallbackParams {
    pub error: Option<String>,
    pub code: Option<String>,
    pub state: Option<String>,
}

#[post("/google_api/authorizations/callback")]
pub async fn callback(
    req: HttpRequest,
    params: web::Form<CallbackParams>,
    session: web::Data<Session>,
) -> impl Responder {
    // Simulate flash messages
    let mut flash_alert = None;
    let mut redirect_uri = redirect_uri_from_session(&params, &session);

    if params.error.is_some() {
        flash_alert = Some("Google Cloud authorizations required".to_string());
        redirect_uri = session.get("error_uri").cloned();
    } else if let Some(code) = &params.code {
        let client = GoogleApiCloudPlatformClient::new(None, "callback_google_api_auth_url");
        let (token, expires_at) = client.get_token(code);
        // Save to session (simulated)
        // session.insert(GoogleApiCloudPlatformClient::session_key_for_token().to_string(), token);
        // session.insert(GoogleApiCloudPlatformClient::session_key_for_expires_at().to_string(), expires_at);
        redirect_uri = redirect_uri_from_session(&params, &session);
    }
    // Simulate Faraday errors as a generic error branch
    // In real code, handle HTTP client errors here
    // Always redirect
    let location = redirect_uri.unwrap_or_else(|| "/".to_string());
    let mut response = HttpResponse::Found().append_header(("Location", location));
    if let Some(alert) = flash_alert {
        response.append_header(("X-Flash-Alert", alert));
    }
    response.finish()
}

fn redirect_uri_from_session(params: &CallbackParams, session: &Session) -> Option<String> {
    if let Some(state) = &params.state {
        session
            .get(&GoogleApiCloudPlatformClient::session_key_for_redirect_uri(
                state,
            ))
            .cloned()
    } else {
        None
    }
}
